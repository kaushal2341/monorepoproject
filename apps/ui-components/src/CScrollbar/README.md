# CScrollbar

Scrollbar component made using `react-custom-scrollbars'` **Scrollbars** component.

To use this component first,

```text
import {CScrollbar} from "@cogent/ui-components";
```
Then,
```jsx harmony
const scrollContent = (
    <>
        <p>Lorem Ipsum is simply dummy text of the printing and typesetting industry.
            Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
            Lorem Ipsum is simply dummy text of the printing and typesetting industry.
            Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
            Lorem Ipsum is simply dummy text of the printing and typesetting industry.
            Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
            Lorem Ipsum is simply dummy text of the printing and typesetting industry.
            Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
            when  Lorem Ipsum is simply dummy text of the printing and typesetting industry.
            Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
            Lorem Ipsum is simply dummy text of the printing and typesetting industry.
            Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
        </p>
    </>);

<CScrollbar id="test" style={{ height: 100}}>
      {scrollContent}
</CScrollbar>
```
## Props of CScrollbar
```text
autoHeight: (Boolean) Enable auto-height mode. (default: false)
            When true container grows with content.
            
autoHeightMax: (Number) Set a maximum height for auto-height mode (default: 200)
            
autoHeightMin: (Number) Set a minimum height for auto-height mode (default: 0)
            
autoHide: (Boolean) Enable auto-hide mode (default: false)
          When true tracks will hide automatically and are only visible while scrolling.
            
autoHideDuration: (Number) Duration for hide animation in ms. (default: 200)

autoHideTimeout:  (Number) Hide delay in ms. (default: 1000)

children: (Node) Children of the component.

className: (String) Apply additional css classes.

hideTracksWhenNotNeeded: (Boolean) Hide tracks (visibility: hidden) when content does not overflow container. (default: false)
            
id: (String) Id of the component.
            
onScroll: (Function) Event handler.

onScrollFrame: (Function) Runs inside the animation frame.
            
onScrollStart: (Function) Called when scrolling starts.

onScrollStop: (Function) Called when scrolling stops.

renderThumbHorizontal: (Function) Horizontal thumb element.(change css of the horizontal scrollbar)

renderThumbVertical: (Function) Vertical thumb element.(change css of the vertical scrollbar)

renderTrackHorizontal: (Function) Horizontal thumb element.(change css of the horizontal scrollbar background)

renderTrackVertical: (Function) Vertical track element.(change css of the vertical scrollbar background)

renderView:  (Function) The area within scrollbar where your content will be rendered.

style: (Object) To add inline css.

thumbMinSize: (Number) Minimal thumb size in px. (default: 30)

thumbSize: (Number) Set a fixed size for thumbs in px.

universal: (Boolean) Enable universal rendering (default: false)

```

## Methods available 
```text
scrollTop(top = 0): scroll to the top value

scrollLeft(left = 0): scroll to the left value

scrollToTop(): scroll to top

scrollToBottom(): scroll to bottom

scrollToLeft(): scroll to left

scrollToRight(): scroll to right

getScrollLeft(): get scrollLeft value

getScrollTop(): get scrollTop value

getScrollWidth(): get scrollWidth value

getScrollHeight(): get scrollHeight value

getClientWidth(): get view client width

getClientHeight(): get view client height

getValues(): get an object with values about the current position.
```

**`Note`**: To use the methods provided by CScrollbar, 
we need to create a ref(callback ref) to the component and simply call the methods using the ref.

Example:
```jsx harmony
import React, {PureComponent} from 'react';
import {CScrollbar} from "@cogent/ui-components";
import {Button} from "react-bootstrap";

class ScrollerComponentTest extends PureComponent {

    scrollToTop = () => {
        this.scrollbarsRef.scrollToTop();
    };

    scrollToBottom = () => {
        this.scrollbarsRef.scrollToBottom();
    };

    render() {
        const scrollContent = (
            <>
                <Button onClick={this.scrollToBottom}>Scroll bottom</Button>
                <p>Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
                    when Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s,
                    when Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when<br/>
                    <Button onClick={this.scrollToTop}>Scroll top</Button>
                </p>
            </>);
        return (
            <CScrollbar
                _ref={c => { this.scrollbarsRef = c }}
                id="test"
                style={{height: 100}}
                className="abc"
                renderTrackVertical={props => <div {...props} className="track-vertical"/>}
                renderThumbVertical={props => <div {...props} className="thumb-vertical"/>}
                renderTrackHorizontal={props => <div {...props} className="track-horizontal"/>}
                renderThumbHorizontal={props => <div {...props} className="thumb-horizontal"/>}
                renderView={props => <div {...props} className="view"/>}
            >
                {scrollContent}
            </CScrollbar>
        );
    }
}
```
**`Note:`** While using the customization props(renderTrackHorizontal, renderThumbHorizontal, renderTrackVertical, 
renderThumbVertical,renderView), we will always need to pass through the given props for the respective element like 
in the example above: <div {...props} className="track-horizontal"/>. This is because we need to pass some default 
styles down to the element in order to make the component work.

# References

-[x] [Malte Wessel](https://malte-wessel.com/react-custom-scrollbars/)
-[x] [Issue fix](https://github.com/malte-wessel/react-custom-scrollbars/issues/119)