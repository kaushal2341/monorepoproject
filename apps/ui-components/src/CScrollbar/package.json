{
  "name": "@cogent/c-scrollbar",
  "version": "0.0.0",
  "main": "CScrollbar.js",
  "author": "Sabu Shakya",
  "license": "MIT",
  "private": true
}
