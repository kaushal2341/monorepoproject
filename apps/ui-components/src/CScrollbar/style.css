.track-vertical {
    top: 2px;
    bottom: 2px;
    right: 2px;
    border-radius: 3px;
    /*background: rgba(255, 255, 255, 0.111);*/
    background:pink;
}
.track-vertical .thumb-vertical {
    position: relative;
    display: block;
    width: 100%;
    cursor: pointer;
    border-radius: inherit;
    /*background-color: rgba(123, 154, 255, 0.418);*/
    background-color: red;
}

.track-horizontal {
    position: absolute;
    height: 6px;
    right: 2px;
    bottom: 2px;
    left: 2px;
    border-radius: 3px;
    background-color: lightpink;
}
.track-horizontal .thumb-horizontal {
    position: relative;
    display: block;
    height: 100%;
    cursor: pointer;
    border-radius: inherit;
    /*background-color: rgba(125, 149, 255, 0.993);*/
    background-color: hotpink;
}

.view {
    position: absolute;
    top: 0px;
    left: 0px;
    right: 0px;
    bottom: 0px;
    overflow: scroll;
    margin-right: -15px;
    margin-bottom: -17px !important;
    background: #FCE1E5;
}

/*SCSS
.track-vertical {
                        top: 2px;
                        bottom: 2px;
                        right: 2px;
                        border-radius: 3px;
                        background: rgba(255, 255, 255, 0.111);

                        .thumb-vertical {
                            position: relative;
                            display: block;
                            width: 100%;
                            cursor: pointer;
                            border-radius: inherit;
                            background-color: rgba(123, 154, 255, 0.418); // changed the color
                        }

                    }

                    .track-horizontal {
                        position: absolute;
                        height: 6px;
                        right: 2px;
                        bottom: 2px;
                        left: 2px;
                        border-radius: 3px;

                        .thumb-horizontal {
                            position: relative;
                            display: block;
                            height: 100%;
                            cursor: pointer;
                            border-radius: inherit;
                            background-color: rgba(125, 149, 255, 0.993); // changed the color
                        }
                    }

                    .view {
                        position: absolute;
                        top: 0px;
                        left: 0px;
                        right: 0px;
                        bottom: 0px;
                        overflow: scroll;
                        margin-right: -15px;
                        margin-bottom: -17px !important; // changed from -15px  (default) to -17px (native scrollbar had a portion visible)
                    }

*/