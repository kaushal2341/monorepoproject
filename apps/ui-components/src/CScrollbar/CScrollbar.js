import React, {PureComponent} from 'react';
import {Scrollbars} from 'react-custom-scrollbars';
import PropTypes from 'prop-types';
import './style.css';

class CScrollbar extends PureComponent {
    render() {
        const {
            autoHeight,
            autoHeightMax,
            autoHeightMin,
            autoHide,
            autoHideDuration,
            autoHideTimeout,
            children,
            className,
            hideTracksWhenNotNeeded,
            id,
            onScroll,
            onScrollFrame,
            onScrollStart,
            onScrollStop,
            onUpdate,
            _ref,
            renderThumbHorizontal,
            renderThumbVertical,
            renderTrackHorizontal,
            renderTrackVertical,
            renderView,
            style,
            thumbMinSize,
            thumbSize,
            universal,
        } = this.props;

        return (
            <Scrollbars
                autoHeight={autoHeight}
                autoHeightMax={autoHeightMax}
                autoHeightMin={autoHeightMin}
                autoHide={autoHide}
                autoHideDuration={autoHideDuration}
                autoHideTimeout={autoHideTimeout}
                children={children}
                className={className}
                hideTracksWhenNotNeeded={hideTracksWhenNotNeeded}
                id={"scrollbar_".concat(id)}
                onScroll={onScroll}
                onScrollFrame={onScrollFrame}
                onScrollStart={onScrollStart}
                onScrollStop={onScrollStop}
                onUpdate={onUpdate}
                ref={_ref}
                renderThumbHorizontal={renderThumbHorizontal}
                renderThumbVertical={renderThumbVertical}
                renderTrackHorizontal={renderTrackHorizontal}
                renderTrackVertical={renderTrackVertical}
                renderView={renderView}
                style={style}
                thumbMinSize={thumbMinSize}
                thumbSize={thumbSize}
                universal={universal}
            />
        );
    }
}

CScrollbar.defaultProps = {
    renderTrackVertical: props => <div {...props} className="track-vertical"/>,
    renderThumbVertical: props => <div {...props} className="thumb-vertical"/>,
    renderTrackHorizontal: props => <div {...props} className="track-horizontal"/>,
    renderThumbHorizontal: props => <div {...props} className="thumb-horizontal"/>,
    renderView: props => <div {...props} className="view"/>,
    thumbSize: 30,
};

CScrollbar.propTypes = {
    autoHeight: PropTypes.bool,
    autoHeightMax: PropTypes.number,
    autoHeightMin: PropTypes.number,
    autoHide: PropTypes.bool,
    autoHideDuration: PropTypes.number,
    autoHideTimeout: PropTypes.number,
    children: PropTypes.node,
    className: PropTypes.string,
    hideTracksWhenNotNeeded: PropTypes.bool,
    id: PropTypes.string.isRequired,
    onScroll: PropTypes.func,
    onScrollFrame: PropTypes.func,
    onScrollStart: PropTypes.func,
    onScrollStop: PropTypes.func,
    onUpdate: PropTypes.func,
    ref: PropTypes.oneOfType([
        PropTypes.func,
        PropTypes.shape({current: PropTypes.instanceOf(Element)})
    ]),
    renderThumbHorizontal: PropTypes.func,
    renderThumbVertical: PropTypes.func,
    renderTrackHorizontal: PropTypes.func,
    renderTrackVertical: PropTypes.func,
    renderView: PropTypes.func,
    style: PropTypes.object,
    thumbMinSize: PropTypes.number,
    thumbSize: PropTypes.number,
    universal: PropTypes.bool,
};

export default CScrollbar;