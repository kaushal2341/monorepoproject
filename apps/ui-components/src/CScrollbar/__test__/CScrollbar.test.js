import CScrollbar from "../CScrollbar";

expect.addSnapshotSerializer(enzymeSerializer);

describe('CScrollbar Tests', () => {
    let sWrapper, mockFunction,mockFunction1;

    beforeEach(() => {
        mockFunction = jest.fn();
        mockFunction1 = jest.fn();
        sWrapper = shallow(<CScrollbar id="test"
                                       onScroll={mockFunction}
                                       onScrollFrame={mockFunction}
                                       onScrollStart={mockFunction}
                                       onScrollStop={mockFunction}
                                       onUpdate={mockFunction1}/>);
    });

    it('should be defined', () => {
        expect(sWrapper.find('CScrollbar')).toBeDefined();
    });

    it('should contain one Scrollbar component', () => {
        expect(sWrapper.find('#scrollbar_test').length).toBe(1);
    });

    it('should have all props available', () => {
        let propsAvailable = [
            'autoHeight',
            'autoHeightMax',
            'autoHeightMin',
            'autoHide',
            'autoHideDuration',
            'autoHideTimeout',
            'children',
            'className',
            'hideTracksWhenNotNeeded',
            'id',
            'onScroll',
            'onScrollFrame',
            'onScrollStart',
            'onScrollStop',
            'onUpdate',
            'renderThumbHorizontal',
            'renderThumbVertical',
            'renderTrackHorizontal',
            'renderTrackVertical',
            'renderView',
            'style',
            'thumbMinSize',
            'thumbSize',
            'universal',
        ];
        let propsOfScrollbar = Object.keys(sWrapper.find('#scrollbar_test').props());
        propsAvailable.forEach((propAvail, i) => (
            expect(propAvail).toContain(propsOfScrollbar[i])
        ));
    });

    it('should call the mockFunction onScroll Event', () => {
        sWrapper.find('#scrollbar_test').simulate('scroll');
        expect(mockFunction).toHaveBeenCalled();
    });

    it('should call the mockFunction onScrollFrame Event', () => {
        sWrapper.find('#scrollbar_test').simulate('scrollFrame');
        expect(mockFunction).toHaveBeenCalled();
    });

    it('should call the mockFunction onScrollStart Event', () => {
        sWrapper.find('#scrollbar_test').simulate('scrollStart');
        expect(mockFunction).toHaveBeenCalled();
    });

    it('should call the mockFunction onScrollStop Event', () => {
        sWrapper.find('#scrollbar_test').simulate('scrollStop');
        expect(mockFunction).toHaveBeenCalled();
    });

    it('should call the mockFunction1 onUpdate', () => {
        sWrapper.find('#scrollbar_test').simulate('update');
        expect(mockFunction1).toHaveBeenCalled();
    });

    describe('Snapshot Testing', () => {
        it('should match snapshot', () => {
            expect(sWrapper).toMatchSnapshot();
        })
    });
});