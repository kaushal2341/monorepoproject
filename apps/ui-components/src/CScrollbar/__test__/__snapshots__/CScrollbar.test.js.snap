// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CScrollbar Tests Snapshot Testing should match snapshot 1`] = `
<Scrollbars
  autoHeight={false}
  autoHeightMax={200}
  autoHeightMin={0}
  autoHide={false}
  autoHideDuration={200}
  autoHideTimeout={1000}
  hideTracksWhenNotNeeded={false}
  id="scrollbar_test"
  onScroll={[MockFunction]}
  onScrollFrame={[MockFunction]}
  onScrollStart={[MockFunction]}
  onScrollStop={[MockFunction]}
  onUpdate={[MockFunction]}
  renderThumbHorizontal={[Function]}
  renderThumbVertical={[Function]}
  renderTrackHorizontal={[Function]}
  renderTrackVertical={[Function]}
  renderView={[Function]}
  tagName="div"
  thumbMinSize={30}
  thumbSize={30}
  universal={false}
/>
`;
